EIPS/eip-7002.md

```diff
@@ -0,0 +1,372 @@
+        new_excess_exits = previous_excess_exits + exit_count - TARGET_EXITS_PER_BLOCK
+    
+    sstore(EXIT_PRECOMPILE_ADDRESS, EXCESS_EXITS_STORAGE_SLOT, new_excess_exits)
+    
+def reset_exit_count():
+    sstore(EXIT_PRECOMPILE_ADDRESS, EXIT_COUNT_STORAGE_SLOT, 0)
+```
+
+#### Block structure
+
+Beginning with the `FORK_BLOCK`, the block body **MUST** be appended with a list of exit operations. RLP encoding of the extended block body structure **MUST** be computed as follows:
+
+```python
+block_body_rlp = RLP([
+    rlp_encoded_field_0,
+    ...,
+    # Latest block body field before `exits`
+    rlp_encoded_field_n,
+    
+    RLP([rlp_encoded_exit_0, ..., rlp_encoded_exit_k]),
+])
+```
+
+Beginning with the `FORK_BLOCK`, the block header **MUST** be appended with the new **`exits_root`** field. The value of this field is the trie root committing to the list of exits in the block body. **`exits_root`** field value **MUST** be computed as follows:
+
+```python
+def compute_trie_root_from_indexed_data(data):
+    trie = Trie.from([(i, obj) for i, obj in enumerate(data)])
+    return trie.root
+
+block.header.exits_root = compute_trie_root_from_indexed_data(block.body.exits)
+```
+
+#### Block validity
+
+Beginning with the `FORK_BLOCK`, client software **MUST** extend block validity rule set with the following conditions:
+
+1. Value of **`exits_root`** block header field equals to the trie root committing to the list of exit operations contained in the block. To illustrate:
+
+```python
+def compute_trie_root_from_indexed_data(data):
+    trie = Trie.from([(i, obj) for i, obj in enumerate(data)])
+    return trie.root
+
+assert block.header.exits_root == compute_trie_root_from_indexed_data(block.body.exits)
+```
+
+2. The list of exit operations contained in the block body **MUST** be equivalent to list of exits at the head of the exit precompile's exit message queue up to the maximum of `MAX_EXITS_PER_BLOCK`, respecting the order in the queue. To illustrate:
+
+```python
+class Exit(object):
+    source_address: Bytes20
+    validator_index: uint64
+
+queue_head_index = sload(EXIT_PRECOMPILE_ADDRESS, EXIT_MESSAGE_QUEUE_HEAD_STORAGE_SLOT)
+queue_tail_index = sload(EXIT_PRECOMPILE_ADDRESS, EXIT_MESSAGE_QUEUE_TAIL_STORAGE_SLOT)
+num_exits_in_queue = queue_tail_index - queue_head_index
+num_exits_to_dequeue = min(num_exits_in_queue, MAX_EXITS_PER_BLOCK)
+
+# Retrieve exits from the queue
+expected_exits = []
+for i in range(num_exits_to_dequeue):
+    queue_storage_slot = EXIT_MESSAGE_QUEUE_STORAGE_OFFSET + (queue_head_index + i) * 2
+    source_address = SLOAD(EXIT_PRECOMPILE_ADDRESS, queue_storage_slot)
+    validator_index = SLOAD(EXIT_PRECOMPILE_ADDRESS, queue_storage_slot + 1)
+    exit = Exit(
+        source_address=Bytes20(source_address),
+        validator_index=uint64(validator_index),
+    )
+    expected_exits.append(exit)
+
+# Compare retrieved exits to the list in the block body
+assert block.body.exits == expected_exits
```

Comment:
couldn't figure out when does a block applies exit in EL (i.e. pops the queue), i am assuming atleast after txs are applied so as to include if there was anything added to the queue

Reply:
I'll make a note to clarify. Yes, after all TXs are applied

* TXs
* check expected exits are correct
* perform precompile computations (move queue pointers, update excess exits, etc)